src/Miso/Internal.hs

```diff
@@ -133,6 +134,248 @@ data ComponentState model action
   , componentActions    :: IORef (Seq action)
   }
 -----------------------------------------------------------------------------
+-- | A 'Topic' represents a place to send and receive messages. 'Topic' is used to facilitate
+-- communication between 'Component'. 'Component' can 'subscribe' to or 'publish' to any 'Topic',
+-- within the same 'Component' or across 'Component'.
+--
+-- This requires creating a custom 'ToJSON' / 'FromJSON'. Any other 'Component'
+-- can 'publish' or 'subscribe' to this @Topic message@. It is a way to provide
+-- loosely-coupled communication between 'Components'.
+--
+-- See 'publish', 'subscribe', 'unsubscribe' for more details.
+--
+-- When distributing 'Component' for third-party use, it is recommended to export
+-- the 'Topic', where 'message' is the JSON protocol.
+--
+--
+-- @since 1.9.0.0
+newtype Topic a = Topic MisoString
+  deriving (Ord, Eq, Show, ToMisoString)
+-----------------------------------------------------------------------------
+-- | Smart constructor for creating a @Topic message@ to write to
+--
+-- @
+--
+-- data Message
+--   = Increment
+--   | Decrement
+--   deriving (Show, Eq, Generic, ToJSON, FromJSON)
+--
+-- arithmetic :: Topic Message
+-- arithmetic = topic "arithmetic"
+--
+-- data Action
+--   = Notification (Result Message)
+--   | Subscribe
+--   | Unsubscribe
+--
+-- update_ :: Action -> Effect Int Action
+-- update_ = \case
+--   Unsubscribe ->
+--     unsubscribe arithmetic
+--   Subscribe ->
+--     subscribe arithmetic Notification
+--   Notification (Success Increment) ->
+--     update_ AddOne
+--   Notification (Success Decrement) ->
+--     update_ SubtractOne
+--   Notification (Error msg) ->
+--     io_ $ consoleError ("Decode failure: " <> ms msg)
+--
+-- @
+--
+-- @since 1.9.0.0
+topic :: MisoString -> Topic a
+topic = Topic
+-----------------------------------------------------------------------------
+mailboxes :: IORef (Map (Topic a) Mailbox)
+{-# NOINLINE mailboxes #-}
+mailboxes = unsafePerformIO $ liftIO (newIORef mempty)
```

Comment:
The use of a phantom type for topics was inspired by #1003, but I noticed when rebasing that there's a difference here in implementation. I had `mailboxes :: IORef (Map MisoString Mailbox)` instead. Using `unsafePerformIO` over a polymorphic reference is seriously dangerous in general, because it makes it easy to accidentally do unsafe coerces, as noted [in the docs](https://hackage.haskell.org/package/base-4.21.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO). But this is internal code and we seem to get away with it, so maybe it doesn't really matter!

Reply:
It's safe here in this case because the underlying type is still `MisoString` (the phantom type isn't coerced at runtime since its erased). In the example you linked [here](https://hackage.haskell.org/package/base-4.21.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO) `Char` is being swapped for `Int`, and this actually no longer core dumps, but it prints `*` which is a coercion yes, but it's a different use case than what's going on in this code.